service/src/ConstConfigIOGroup.cpp

```diff
@@ -0,0 +1,451 @@
+                                           const std::string &default_file_path)
+    {
+        std::string config_json;
+        bool load_default = true;
+        if (!user_file_path.empty()) {
+            try {
+                config_json = geopm::read_file(user_file_path);
+                load_default = false;
+            } catch (...) {
+#ifdef GEOPM_DEBUG
+                std::cerr << "Warning: <geopm> Failed to load "
+                          << "ConstConfigIOGroup configuration file: "
+                          << user_file_path
+                          << ". Proceeding with default configuration file..."
+                          << std::endl;
+#endif
+            }
+        }
+
+        if (load_default) {
+            config_json = geopm::read_file(default_file_path);
+        }
+
+        parse_config_json(config_json);
+    }
+
+    std::set<std::string> ConstConfigIOGroup::signal_names(void) const
+    {
+        std::set<std::string> result;
+        for (const auto &sv : m_signal_available) {
+            result.insert(sv.first);
+        }
+        return result;
+    }
+
+    std::set<std::string> ConstConfigIOGroup::control_names(void) const
+    {
+        return {};
+    }
+
+    bool ConstConfigIOGroup::is_valid_signal(const std::string &signal_name) const
+    {
+        return m_signal_available.find(signal_name) != m_signal_available.end();
+    }
+
+    bool ConstConfigIOGroup::is_valid_control(const std::string &control_name) const
+    {
+        return false;
+    }
+
+    int ConstConfigIOGroup::signal_domain_type(const std::string &signal_name) const
+    {
+        int result = GEOPM_DOMAIN_INVALID;
+        auto it = m_signal_available.find(signal_name);
+
+        if (it != m_signal_available.end())
+            result = it->second->domain;
+
+        return result;
+    }
+
+    int ConstConfigIOGroup::control_domain_type(const std::string &control_name) const
+    {
+        return GEOPM_DOMAIN_INVALID;
+    }
+
+    int ConstConfigIOGroup::push_signal(const std::string &signal_name,
+                                        int domain_type, int domain_idx)
+    {
+        auto it = m_signal_available.find(signal_name);
+        if (it == m_signal_available.end()) {
+            throw Exception("ConstConfigIOGroup::push_signal(): " +
+                            signal_name + " not valid for ConstConfigIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        else if (domain_type != it->second->domain) {
+            throw Exception("ConstConfigIOGroup::push_signal(): domain_type " +
+                            std::to_string(domain_type) +
+                            " not valid for ConstConfigIOGroup",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        else if (domain_idx < 0 ||
+            static_cast<std::size_t>(domain_idx) >= it->second->values.size()) {
```

Comment:
Is there a reason why this would be preferred? I'm thinking casting ``int`` to ``unsigned int`` (after we've ensured it's >= 0) is safer than casting ``unsigned int`` to ``int``. I highly doubt the ``values`` vector would ever be so large for that to matter (under normal use cases), but I would still prefer to keep do the safer cast (in case of malicious input).

Reply:
Casting a signed int to an unsigned int can result in weird values if the initial value was negative (e.g. 2^32 - 1 or similar for an initial value of -1).  It's why we avoid this sort of thing in other spots in the code.

Casting an unsigned int to a signed int should always be safe unless there's a difference in precision that results in data truncation.